test/hotspot/jtreg/compiler/c2/irTests/TestIfMinMax.java

```diff
@@ -0,0 +1,188 @@
+import jdk.test.lib.Utils;
+
+/*
+ * @test
+ * @bug 8324655
+ * @summary Test that if expressions are properly folded into min/max nodes
+ * @library /test/lib /
+ * @run driver compiler.c2.irTests.TestIfMinMax
+ */
+public class TestIfMinMax {
+    private static final Random RANDOM = Utils.getRandomInstance();
+
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.MIN_I, "1" })
+    public int testMinI1(int a, int b) {
+        return a < b ? a : b;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.MIN_I, "1" })
+    public int testMinI2(int a, int b) {
+        return a > b ? b : a;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.MAX_I, "1" })
+    public int testMaxI1(int a, int b) {
+        return a > b ? a : b;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.MAX_I, "1" })
+    public int testMaxI2(int a, int b) {
+        return a < b ? b : a;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.MIN_I, "1" })
+    public int testMinI1E(int a, int b) {
+        return a <= b ? a : b;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.MIN_I, "1" })
+    public int testMinI2E(int a, int b) {
+        return a >= b ? b : a;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.MAX_I, "1" })
+    public int testMaxI1E(int a, int b) {
+        return a >= b ? a : b;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.MAX_I, "1" })
+    public int testMaxI2E(int a, int b) {
+        return a <= b ? b : a;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.CMOVE_L, "1" })
+    public long testMinL1(long a, long b) {
+        return a < b ? a : b;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.CMOVE_L, "1" })
+    public long testMinL2(long a, long b) {
+        return a > b ? b : a;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.CMOVE_L, "1" })
+    public long testMaxL1(long a, long b) {
+        return a > b ? a : b;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.CMOVE_L, "1" })
+    public long testMaxL2(long a, long b) {
+        return a < b ? b : a;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.CMOVE_L, "1" })
+    public long testMinL1E(long a, long b) {
+        return a <= b ? a : b;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.CMOVE_L, "1" })
+    public long testMinL2E(long a, long b) {
+        return a >= b ? b : a;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.CMOVE_L, "1" })
+    public long testMaxL1E(long a, long b) {
+        return a >= b ? a : b;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.IF }, counts = { IRNode.CMOVE_L, "1" })
+    public long testMaxL2E(long a, long b) {
+        return a <= b ? b : a;
+    }
```

Comment:
I assume some of the `long` patterns should also have become MaxL/MinL in some phase, right? Is there maybe some phase where the IR would actually show that? You can target the IR rule to a phase, I think. Would be worth a try.

Reply:
Oh true, I think we can identify MinL/MaxL before macro expansion is done.